From b6fd351bd17180914106422bf6b1074ac44be242 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@gmail.com>
Date: Fri, 25 Sep 2026 04:27:32 -0400
Subject: [PATCH] Add board BananaPi BPI-M4-Zero rev 2

- Rev2 (Rayson RS1G32LX4 D4BNR-53BT2404)

Adjust LPDDR4 timings to stabilize 792 MHz operation

The current LPDDR4 parameters for the H616 platform fail at 792 MHz,
leading to data corruption ("Bad Data CRC") and "Synchronous Aborts"
during early boot phases. This instability is driven by excessive
signal drive strengths causing trace ringing and misaligned PHY data lane
sampling windows at higher frequencies.

Stabilize 792 MHz operation by tuning TPR6 and TPR11:

- CONFIG_DRAM_SUNXI_TPR6: Attenuated from 0x48808080 to 0x38808080.
  This scales back the output driver impedance to mitigate high-frequency
  signal reflections and ringing on the PCB layout.

- CONFIG_DRAM_SUNXI_TPR11: Adjusted data lane skews from 0x26262524
  to 0x25252624. This shifts the internal sampling strobe, compensating
  for trace length variations and ensuring all 32 bits are captured
  squarely within the stable center of the data eye.

Retain deep FIFO buffers (TPR12=0x100f100f) to handle the 792 MHz data rate
without risking underflow.

Signed-off-by: Patrick Yavitz <pyavitz@gmail.com>
---
 configs/bananapi_m4zero_defconfig             |  32 +++
 .../sun50i-h618-bananapi-m4-zero-rev-2.dts    |  40 +++
 .../sun50i-h618-bananapi-m4-zero.dts          |  12 +
 .../allwinner/sun50i-h618-bananapi-m4.dtsi    | 248 ++++++++++++++++++
 4 files changed, 332 insertions(+)
 create mode 100644 configs/bananapi_m4zero_defconfig
 create mode 100644 dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero-rev-2.dts
 create mode 100644 dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero.dts
 create mode 100644 dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4.dtsi

diff --git a/configs/bananapi_m4zero_defconfig b/configs/bananapi_m4zero_defconfig
new file mode 100644
index 00000000000..c0b8bb20bd2
--- /dev/null
+++ b/configs/bananapi_m4zero_defconfig
@@ -0,0 +1,32 @@
+CONFIG_OF_UPSTREAM=y
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_DEFAULT_DEVICE_TREE="allwinner/sun50i-h618-bananapi-m4-zero"
+CONFIG_SPL=y
+CONFIG_DRAM_SUNXI_DX_ODT=0x07070707
+CONFIG_DRAM_SUNXI_DX_DRI=0x0e0e0e0e
+CONFIG_DRAM_SUNXI_CA_DRI=0x0e0e
+CONFIG_DRAM_SUNXI_ODT_EN=0xaaaaeeee
+CONFIG_DRAM_SUNXI_TPR6=0x38808080
+CONFIG_DRAM_SUNXI_TPR10=0x402f6663
+CONFIG_DRAM_SUNXI_TPR11=0x25252624
+CONFIG_DRAM_SUNXI_TPR12=0x100f100f
+CONFIG_MACH_SUN50I_H616=y
+CONFIG_SUNXI_DRAM_H616_LPDDR4=y
+CONFIG_DRAM_CLK=792
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
+CONFIG_R_I2C_ENABLE=y
+# CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
+CONFIG_CMD_KASLRSEED=y
+CONFIG_SPL_I2C=y
+CONFIG_SPL_SYS_I2C_LEGACY=y
+CONFIG_SYS_I2C_MVTWSI=y
+CONFIG_SYS_I2C_SLAVE=0x7f
+CONFIG_SYS_I2C_SPEED=400000
+CONFIG_SUN8I_EMAC=y
+CONFIG_SUPPORT_EMMC_BOOT=y
+CONFIG_AXP313_POWER=y
+CONFIG_DM_RNG=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_MUSB_GADGET=y
diff --git a/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero-rev-2.dts b/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero-rev-2.dts
new file mode 100644
index 00000000000..0b55e4fd887
--- /dev/null
+++ b/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero-rev-2.dts
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2026 Patrick Yavitz <pyavitz@gmail.com>
+ */
+
+/dts-v1/;
+
+#include "sun50i-h618-bananapi-m4.dtsi"
+
+/ {
+	model = "BananaPi BPI-M4-Zero rev.2";
+};
+
+/* SDIO */
+&mmc1 {
+	status = "okay";
+
+	sdio: wifi@1 {
+		reg = <1>;
+		compatible = "brcm,bcm4329-fmac";
+	};
+};
+
+/* Bluetooth */
+&uart1 {
+	status = "okay";
+	pinctrl-0 = <&uart1_pins>, <&uart1_rts_cts_pins>;
+	pinctrl-names = "default";
+	uart-has-rtscts;
+
+	bluetooth {
+		compatible = "brcm,bcm43540-bt";
+		max-speed = <1500000>;
+		host-wakeup-gpios = <&pio 6 16 GPIO_ACTIVE_HIGH>;
+		device-wakeup-gpios = <&pio 6 17 GPIO_ACTIVE_HIGH>;
+		shutdown-gpios = <&pio 6 19 GPIO_ACTIVE_HIGH>;
+		vbat-supply = <&reg_vcc3v3>;
+		vddio-supply = <&reg_vcc1v8>;
+	};
+};
diff --git a/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero.dts b/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero.dts
new file mode 100644
index 00000000000..574eff97c44
--- /dev/null
+++ b/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4-zero.dts
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2026 Patrick Yavitz <pyavitz@gmail.com>
+ */
+
+/dts-v1/;
+
+#include "sun50i-h618-bananapi-m4.dtsi"
+
+/ {
+	model = "BananaPi BPI-M4-Zero";
+};
diff --git a/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4.dtsi b/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4.dtsi
new file mode 100644
index 00000000000..a2cf593244b
--- /dev/null
+++ b/dts/upstream/src/arm64/allwinner/sun50i-h618-bananapi-m4.dtsi
@@ -0,0 +1,248 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2026 Patrick Yavitz <pyavitz@gmail.com>
+ */
+
+/dts-v1/;
+
+#include "sun50i-h616.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/linux-event-codes.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	compatible = "sinovoip,bpi-m4-zero", "allwinner,sun50i-h618";
+
+	aliases {
+		serial0 = &uart0;
+		serial5 = &uart5;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-0 {
+			color = <LED_COLOR_ID_RED>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&pio 2 12 GPIO_ACTIVE_HIGH>;	/* PC12 */
+			linux,default-trigger = "heartbeat";
+		};
+	};
+
+	reg_usb_vbus: regulator-usb-vbus {
+		/* Separate discrete regulator for the USB ports */
+		compatible = "regulator-fixed";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-name = "usb-vbus";
+		vin-supply = <&reg_vcc5v>;
+	};
+
+	reg_vcc5v: regulator-vcc5v {
+		/* Board wide 5V supply directly from the USB-C socket */
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		regulator-name = "vcc-5v";
+	};
+
+	reg_vcc3v3: regulator-vcc3v3 {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-name = "vcc-3v3";
+		vin-supply = <&reg_vcc5v>;
+	};
+
+	reg_vcc1v8: regulator-vcc1v8 {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		regulator-name = "vcc-1v8";
+		vin-supply = <&reg_vcc3v3>;
+	};
+
+	wifi_pwrseq: wifi-pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		clocks = <&rtc CLK_OSC32K_FANOUT>;
+		clock-names = "ext_clock";
+		pinctrl-0 = <&x32clk_fanout_pin>;
+		pinctrl-names = "default";
+		post-power-on-delay-ms = <200>;
+		reset-gpios = <&pio 6 18 GPIO_ACTIVE_LOW>; /* PG18 */
+	};
+};
+
+&cpu0 {
+	cpu-supply = <&reg_dcdc2>;
+};
+
+/* Connected to an on-board RTL8821CU USB WiFi chip. */
+&ehci1 {
+	status = "disabled";
+};
+
+&ehci3 {
+	status = "okay";
+};
+
+/* SD card */
+&mmc0 {
+	status = "okay";
+	bus-width = <4>;
+	max-frequency = <50000000>;
+	disable-wp;
+	cd-gpios = <&pio 5 6 GPIO_ACTIVE_HIGH>;	/* PF6 */
+	vmmc-supply = <&reg_vcc3v3>;
+};
+
+/* SDIO */
+&mmc1 {
+	status = "disabled";
+	bus-width = <4>;
+	max-frequency = <100000000>;
+	non-removable;
+	disable-wp;
+	/* WiFi firmware requires power to be kept while in suspend */
+	keep-power-in-suspend;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	cd-gpios = <&pio 6 15 GPIO_ACTIVE_HIGH>; /* PG15 */
+	vmmc-supply = <&reg_vcc3v3>;
+};
+
+/* eMMC */
+&mmc2 {
+	status = "okay";
+	bus-width = <8>;
+	cap-mmc-hw-reset;
+	mmc-hs200-1_8v;
+	non-removable;
+	disable-wp;
+	vmmc-supply = <&reg_vcc3v3>;
+	vqmmc-supply = <&reg_vcc1v8>;
+};
+
+&ohci3 {
+	status = "okay";
+};
+
+&pio {
+	vcc-pc-supply = <&reg_aldo1>;
+	vcc-pf-supply = <&reg_dldo1>;
+	vcc-pg-supply = <&reg_dldo1>;
+	vcc-ph-supply = <&reg_dldo1>;
+	vcc-pi-supply = <&reg_dldo1>;
+
+	/* Add UART Pins */
+	uart4_pi_pins: uart4-pi-pins {
+		pins = "PI13", "PI14";
+		function = "uart4";
+	};
+
+	uart4_pi_rts_cts_pins: uart4-pi-rts-cts-pins {
+		pins = "PI15", "PI16";
+		function = "uart4";
+	};
+
+	uart5_ph_pins: uart5-ph-pins {
+		pins = "PH2", "PH3";
+		function = "uart5";
+	};
+};
+
+&r_i2c {
+	status = "okay";
+
+	axp313: pmic@36 {
+		compatible = "x-powers,axp313a";
+		reg = <0x36>;
+		#interrupt-cells = <1>;
+		interrupt-controller;
+		interrupt-parent = <&pio>;
+
+		vin1-supply = <&reg_vcc5v>;
+		vin2-supply = <&reg_vcc5v>;
+		vin3-supply = <&reg_vcc5v>;
+
+		regulators {
+			reg_aldo1: aldo1 {
+				regulator-always-on;
+				regulator-min-microvolt = <1800000>;
+				regulator-max-microvolt = <1800000>;
+				regulator-name = "vcc-1v8-pll";
+			};
+
+			reg_dldo1: dldo1 {
+				regulator-always-on;
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-name = "vcc-3v3-io";
+			};
+
+			reg_dcdc1: dcdc1 {
+				regulator-always-on;
+				regulator-min-microvolt = <810000>;
+				regulator-max-microvolt = <990000>;
+				regulator-name = "vdd-gpu-sys";
+			};
+
+			reg_dcdc2: dcdc2 {
+				regulator-always-on;
+				regulator-min-microvolt = <810000>;
+				regulator-max-microvolt = <1100000>;
+				regulator-name = "vdd-cpu";
+			};
+
+			reg_dcdc3: dcdc3 {
+				regulator-always-on;
+				regulator-min-microvolt = <1100000>;
+				regulator-max-microvolt = <1100000>;
+				regulator-name = "vdd-dram";
+			};
+		};
+	};
+};
+
+&uart0 {
+	status = "okay";
+	pinctrl-0 = <&uart0_ph_pins>;
+	pinctrl-names = "default";
+};
+
+/* Bluetooth */
+&uart1 {
+	status = "disabled";
+	pinctrl-0 = <&uart1_pins>;
+	pinctrl-names = "default";
+};
+
+&uart4 {
+	status = "disabled";
+	pinctrl-0 = <&uart4_pi_pins>;
+	pinctrl-names = "default";
+};
+
+&uart5 {
+	status = "okay";
+	pinctrl-0 = <&uart5_ph_pins>;
+	pinctrl-names = "default";
+};
+
+&usbotg {
+	status = "okay";
+	dr_mode = "peripheral";
+};
+
+&usbphy {
+	status = "okay";
+	usb1_vbus-supply = <&reg_usb_vbus>;
+};
-- 
2.53.0

